### PR TITLE
BRIN fix load/unload with locks, fixes #820

### LINKED ISSUE #820: BRIN page unload deadlock
**BUG REPORT**

1. Please describe the issue you observed:
I found tablespaces stuck on checkpoint procedure. Deep looking on logs and stacks I found that HerdDB was stuck in a deadlock case between a BRIN index value deletion and another BRIN index checkpoint:
Checkpointing thread
```
"hdb-pecora-activator" #38 daemon prio=5 os_prio=0 cpu=6524841.27ms elapsed=88200.79s tid=0x00007f4b66096840 nid=0x1a1065 waiting on condition  [0x00007f2be12fe000]
   java.lang.Thread.State: WAITING (parking)
        at jdk.internal.misc.Unsafe.park(java.base@17.0.9/Native Method)
        - parking to wait for  <0x00007f2e5b4fb468> (a java.util.concurrent.locks.ReentrantLock$FairSync)
        at java.util.concurrent.locks.LockSupport.park(java.base@17.0.9/LockSupport.java:211)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@17.0.9/AbstractQueuedSynchronizer.java:715)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@17.0.9/AbstractQueuedSynchronizer.java:938)
        at java.util.concurrent.locks.ReentrantLock$Sync.lock(java.base@17.0.9/ReentrantLock.java:153)
        at java.util.concurrent.locks.ReentrantLock.lock(java.base@17.0.9/ReentrantLock.java:322)
        at herddb.index.brin.BlockRangeIndex$Block.unload(BlockRangeIndex.java:700)
        at herddb.index.brin.BlockRangeIndex$Block.unload(BlockRangeIndex.java:730)
        at herddb.index.brin.BlockRangeIndex.merge(BlockRangeIndex.java:795)
        at herddb.index.brin.BlockRangeIndex.checkpoint(BlockRangeIndex.java:914)
        at herddb.index.brin.BRINIndexManager.checkpoint(BRINIndexManager.java:317)
        at herddb.core.TableManager.checkpoint(TableManager.java:3253)
        at herddb.core.TableManager.checkpoint(TableManager.java:2670)
        at herddb.core.TableSpaceManager.checkpoint(TableSpaceManager.java:2076)
        at herddb.core.DBManager.checkpoint(DBManager.java:964)
        at herddb.core.DBManager.executeActivator(DBManager.java:1276)
        at herddb.core.DBManager.access$500(DBManager.java:124)
        at herddb.core.DBManager$Activator.run(DBManager.java:1208)
        at java.lang.Thread.run(java.base@17.0.9/Thread.java:840)

   Locked ownable synchronizers:
        - <0x00007f2d51f48b20> (a java.util.concurrent.locks.ReentrantLock$FairSync)
        - <0x00007f2d51f48b88> (a java.util.concurrent.locks.ReentrantLock$FairSync)
        - <0x00007f2d51f48bf0> (a java.util.concurrent.locks.ReentrantLock$FairSync)
```

Another index value deletion thread:
```
"db-dmlcall-localhost:7000-3" #265 prio=5 os_prio=0 cpu=768760.12ms elapsed=88704.87s tid=0x00007f28f0052030 nid=0x1a1226 waiting on condition  [0x00007f27881f8000]
   java.lang.Thread.State: WAITING (parking)
        at jdk.internal.misc.Unsafe.park(java.base@17.0.9/Native Method)
        - parking to wait for  <0x00007f2d51f48b88> (a java.util.concurrent.locks.ReentrantLock$FairSync)
        at java.util.concurrent.locks.LockSupport.park(java.base@17.0.9/LockSupport.java:211)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@17.0.9/AbstractQueuedSynchronizer.java:715)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer.acquire(java.base@17.0.9/AbstractQueuedSynchronizer.java:938)
        at java.util.concurrent.locks.ReentrantLock$Sync.lock(java.base@17.0.9/ReentrantLock.java:153)
        at java.util.concurrent.locks.ReentrantLock.lock(java.base@17.0.9/ReentrantLock.java:322)
        at herddb.index.brin.BlockRangeIndex$Block.unload(BlockRangeIndex.java:700)
        at herddb.index.brin.BlockRangeIndex$Block.unload(BlockRangeIndex.java:730)
        at herddb.index.brin.BlockRangeIndex$Block.ensureBlockLoaded(BlockRangeIndex.java:499)
        at herddb.index.brin.BlockRangeIndex$Block.delete(BlockRangeIndex.java:380)
        at herddb.index.brin.BlockRangeIndex.delete(BlockRangeIndex.java:1055)
        at herddb.index.brin.BRINIndexManager.recordDeleted(BRINIndexManager.java:431)
        at herddb.core.TableManager.applyDelete(TableManager.java:2159)
        at herddb.core.TableManager.onTransactionCommit(TableManager.java:1948)
        at herddb.core.TableSpaceManager.apply(TableSpaceManager.java:428)
        at herddb.core.TableSpaceManager.lambda$commitTransaction$22(TableSpaceManager.java:2188)
        at herddb.core.TableSpaceManager$$Lambda$1815/0x00007f2befd5f7e8.apply(Unknown Source)
        at java.util.concurrent.CompletableFuture.uniHandle(java.base@17.0.9/CompletableFuture.java:934)
        at java.util.concurrent.CompletableFuture$UniHandle.tryFire(java.base@17.0.9/CompletableFuture.java:911)
        at java.util.concurrent.CompletableFuture$Completion.run(java.base@17.0.9/CompletableFuture.java:482)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@17.0.9/ThreadPoolExecutor.java:1136)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@17.0.9/ThreadPoolExecutor.java:635)
        at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(java.base@17.0.9/Thread.java:840)

   Locked ownable synchronizers:
        - <0x00007f2d56d06508> (a java.util.concurrent.ThreadPoolExecutor$Worker)
        - <0x00007f2e5b4fb468> (a java.util.concurrent.locks.ReentrantLock$FairSync)
```

### PR DESCRIPTION
**Pull request by diegosalvi**
This PR fixes #820 
Changes load/unload BRIN pages deferring unload until the unlock of currently handled BRIN block, doing so we don't keep current block locked until other block has been fully unload (that needs a lock too).
Changed block loading during blocks merge too: now we don't load in the page memory blocks that aren't needed anymore because the will deleted at merge end.
The neat result is that there is one more page in thread memory that will discarded/cleaned at the end.

Following this checklist to help us incorporate your 
contribution quickly and easily:

 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [X] Rember to add the correct license header to new files.
 - [X] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will 
       be performed on your pull request automatically.

To make clear that you license your contribution under 
the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)
you have to acknowledge this by using the following check-box.

 - [X ] I hereby declare this contribution to be licenced under the [Apache License Version 2.0, January 2004](http://www.apache.org/licenses/LICENSE-2.0)

### CODE REVIEW

**Comment by eolivelli** [comment]
As we have a trace for the deadlock, would it be possible to add a test case the reproduces the problem ?

**Comment by diegosalvi** [comment]
@eolivelli it had my first concern but I haven't been able do reproduce in a test 'till now

**Comment by diegosalvi** [comment]
Do not integrate! I found some strange behaviours while attempting to reproduce the deadlock.

**Comment by diegosalvi** [comment]
Added a testcase that reproduced the deadlock

**Comment by diegosalvi** [comment]
Now can be merged

**Comment by eolivelli** [comment]
The patch is not trivial, I will take a look this week

Thanks

**Comment by diegosalvi** [comment]
@eolivelli Yes I had to change many things to make it working. There were many bugs in page and lock handling.
Biggest changes are:

1. **Fix load/unload deadlock.** When loading a page the page that should be unloaded (due to page replacement policy) isn't unloaded as fast as possible but only when lock are released. This avoid some "random" deadlock on load&unload. The net effect is that we have one (possibly full) page in memory for the current thread. This can be improved adding a "try unload" mechanic to pages (so... try unload if you can take the lock, abort unload and signal otherwise) but it should be supported by all pages types (brin, blink and datapages): the patch would be too big. We'll reserve the improvement for the future.
2. **Fix phantom node modifications.** Blocks now have a new flag "detached". It is set only during merging to signal that a node block is detached from the tree. We need to have this information somehow due to concurrent search, add and delete: they keep a pointer to a block that can be merged and removed from the tree between a concurrent node find and node lock. Without such information we could remove or add data to a dead node. We could avoid the flag doing a lookup on the tree after lock checking if the node is still there but it is much more expensive
3. **Avoid deleted node checkpoint.** During merge we checkpointed merged node too. It isn't really necessary: they are dead node that are discarded from the tree, they aren't needed anymore, no need to checkpoint them.
4. **Fix add/delete/search with checkpoint deadlock.** Actually is more search with checkpoint deadlock because we don't execute add or deletes concurrently with checkpoints but one change fixes them all. Checkpoint procedure order has been reversed. Before the patch we started from the tail looking for nodes to merge, now we start from the head. Metadata are still writtend in reverse order to be back compatible with existing metadata (and do not rewrite recover procedure too, procedure that is simpler written in a reverse order). Reversing checkpoint procedure order every lock on the tree is taken in the same direction thus avoiding deadlock.

**Comment by eolivelli** [comment]
it would be great to run the long running performance tests with this change before cutting a release.
nasty things may happen in the long run

File: herddb-core/src/main/java/herddb/index/brin/BlockRangeIndex.java [hunk 1]
```diff
@@ -738,7 +805,9 @@ public void unload(long pageId) {
 
         @Override
         public String toString() {
-            return "Block{" + "key=" + key + ", minKey=" + key.minKey + ", size=" + size + ", next=" + (next == null ? null : next.key) + "}";
+            return System.identityHashCode(this) + "Block{" + "key=" + key + ", minKey=" + key.minKey
```
Comment (eolivelli) [inline herddb-core/src/main/java/herddb/index/brin/BlockRangeIndex.java:808]: is identityHashCode a left over ? 
 if you want to keep it we can put it inside the curly braces

**Review by eolivelli — COMMENTED** [review]
I did a first quick look.

I am surprised that we are touching so many method but we are adding only one test case.
Is there any corner case that we are touching that is not covered by tests ?

I am asking because we are touching a critical part of the system

File: herddb-core/src/main/java/herddb/index/brin/BlockRangeIndex.java [hunk 2]
```diff
@@ -316,8 +326,25 @@ void addValue(Key key, Val value, PutState<Key, Val> state) {
...
 
+                if (detached) {
+                    /*
+                     * If this block has been detached it means that current block has been merged by another
+                     * thread before beeing able to lock the block.
+                     * The should be now in the preceeding node and this shouldn't be used anymore.
+                     * Normally we scan the index using the next block reference and we cannot backtrack to previous
+                     * nodes so we need to start another search on the index itself to look for the current data node
+                     */
+                    final boolean fineEnabled = LOG.isLoggable(Level.FINE);
```
[the remark this answers is not in the document]
Comment (diegosalvi) [reply herddb-core/src/main/java/herddb/index/brin/BlockRangeIndex.java:341]: Fine logging are someway bigger in BRIN so we avoid to build strings at all checking on this variable. In this specific case there is only one log (in other part of BRIN there are more fine log lines per method) but I left the variable defined ad in other methods for reading clarity

**Comment by diegosalvi** [comment]
> I did a first quick look.
> 
> I am surprised that we are touching so many method but we are adding only one test case. Is there any corner case that we are touching that is not covered by tests ?
> 
> I am asking because we are touching a critical part of the system

There is only one test case to force the old deadlock situation. Other cases are covered by existing tests (and failed many times during development). The biggest change, the checkpoint order change, is covered by a whole test suite about chekpointing (and testing metadata ordering and produced tree) and restoring/restarting.

**Review by eolivelli — APPROVED** [review]
LGTM

great work

As I said above, I would feel more comfortable to cut a release including this change only after running some long running tests under load

@aluccaroni @hamadodene @amitvc

**Comment by eolivelli** [comment]
@dmercuriali @aluccaroni  @diegosalvi  we can merge this patch.
My point is about not cutting a release before running appropriate load testing.

**Comment by diegosalvi** [comment]
Currently I'm running a test instance with some load, let's see in some days if it doesn't result in some errors

**Comment by diegosalvi** [comment]
With @hamadodene we have run a working instance for a week with load without any issue.

**Comment by eolivelli** [comment]
Great work!
